jit-compiler/src/lib.rs

```diff
@@ -1,22 +1,36 @@
 mod codegen;
 mod compiler;
 
-use std::{collections::HashMap, fs};
+use std::{collections::HashMap, fs, sync::Arc};
 
 use codegen::CodeGenerator;
 use compiler::{call_cargo, generate_glue_code, load_library};
+
 use powdr_ast::analyzed::Analyzed;
 use powdr_number::FieldElement;
 
-pub type SymbolMap = HashMap<String, fn(u64) -> u64>;
+/// Wrapper around a dynamically loaded function.
+/// Prevents the dynamically loaded library to be unloaded while the function is still in use.
+#[derive(Clone)]
+pub struct LoadedFunction {
+    #[allow(dead_code)]
+    library: Arc<libloading::Library>,
+    function: unsafe extern "C" fn(u64) -> u64,
+}
```

Comment:
How about this instead of using `Arc`?
```
pub struct SymbolMap {
    library: libloading::Library
}
```

and then:
```
impl SymbolMap {
    fn get<'a>(&'a self, name: &str) -> Result<libloading::Symbol<'a, extern "C" fn(u64) -> u64>>
    {
        // most of the code currently inside `load_library()`,
        // i.e. the code to get the function
    }
}
```

Looks more elegant, but I can think of two drawbacks:

- The error checking for symbol availability is deferred until `get()` (which should not be a problem if we are loading the library we just built, as the symbol is expected to be there);
- The symbol search is linear (which should not be a problem if each function is retrieved just a few times).

Reply:
According to ChatGPT, symbol lookup is also indexed on macOS and Windows.

Reply:
I don't really want to leak the weird Symbol thing with it's lifetime parameter. In the tests, we are retrieving a single function and it's much more fun to just store or return it and don't care about lifetime.

Reply:
Also what are the issues about Arc? It's a tiny atomic update for each clone / drop, right? I don't think they should happen often.

Reply:
There is no issue with Arc. If you prefer this way, as long as the raw function pointer doesn't leak out of the struct, this is fine. One extra pointer per function won't make any difference.